mobile_verifier/src/speedtests_average.rs

```diff
@@ -0,0 +1,478 @@
+{
+    fn from(iter: &'a I) -> Self {
+        let mut id = vec![]; // eww!
+        let mut window_size = 0;
+        let mut sum_upload = 0;
+        let mut sum_download = 0;
+        let mut sum_latency = 0;
+
+        for Speedtest { report, .. } in
+            speedtests_without_lapsed(iter.into_iter(), Duration::hours(SPEEDTEST_LAPSE))
+        {
+            id = report.pubkey.as_ref().to_vec(); // eww!
+            sum_upload += report.upload_speed;
+            sum_download += report.download_speed;
+            sum_latency += report.latency;
+            window_size += 1;
+        }
+
+        if window_size > 0 {
+            let upload_speed_avg_bps = sum_upload / window_size as u64;
+            let download_speed_avg_bps = sum_download / window_size as u64;
+            let latency_avg_ms = sum_latency / window_size as u32;
+            let validity = validity(
+                window_size as usize,
+                upload_speed_avg_bps,
+                download_speed_avg_bps,
+                latency_avg_ms,
+            );
+            let tier = SpeedtestTier::new(
+                window_size as usize,
+                upload_speed_avg_bps,
+                download_speed_avg_bps,
+                latency_avg_ms,
+            );
+            let reward_multiplier = tier.into_multiplier();
+            SpeedtestAverage {
+                pubkey: id.into(),
+                window_size: window_size as usize,
+                upload_speed_avg_bps,
+                download_speed_avg_bps,
+                latency_avg_ms,
+                validity,
+                reward_multiplier,
+            }
+        } else {
+            SpeedtestAverage {
+                pubkey: id.into(),
+                window_size: 0,
+                upload_speed_avg_bps: sum_upload,
+                download_speed_avg_bps: sum_download,
+                latency_avg_ms: sum_latency,
+                validity: proto::SpeedtestAvgValidity::TooFewSamples,
+                reward_multiplier: Decimal::ZERO,
+            }
+        }
+    }
+}
+
+impl SpeedtestAverage {
+    pub async fn write(
+        &self,
+        filesink: &FileSinkClient,
+        speedtests: Vec<Speedtest>,
+    ) -> file_store::Result {
+        filesink
+            .write(
+                proto::SpeedtestAvg {
+                    pub_key: self.pubkey.clone().into(),
+                    upload_speed_avg_bps: self.upload_speed_avg_bps,
+                    download_speed_avg_bps: self.download_speed_avg_bps,
+                    latency_avg_ms: self.latency_avg_ms,
+                    timestamp: Utc::now().encode_timestamp(),
+                    speedtests: speedtests_without_lapsed(
+                        speedtests.iter(),
+                        Duration::hours(SPEEDTEST_LAPSE),
+                    )
+                    .map(|st| proto::Speedtest {
+                        timestamp: st.report.timestamp.timestamp() as u64,
```

Comment:
This should probably implement the timestamp_millis feature so you don’t have to cast it to a u64, right?

Reply:
copy pasted from the original speedtests module but changed now to use the timestamp trait.  Returns seconds tho not millis

Reply:
addressed here: 67effc3558d2cd2c56ef07956dba0b19b02ffd95